Feature/reduce detail requests

Belal-Alkhatib opened the pull request:
# Pull Request Template

## Description
This PR significantly improves the performance, readability, and maintainability of the movie and TV show details features by consolidating multiple network requests into a single call using append_to_response. It introduces a centralized response handling and domain model mapping strategy.
---

## Changes Made
List the changes introduced in this pull request:
### Movie Details
**API Integration:**
- Switched to single endpoint call using:` /movie/{movie_id}?append_to_response=credits,reviews,similar,images`
- Removed redundant endpoints and methods from MovieApiService.

**Domain + Data Refactor:**
- Added RemoteMovieDetailsResponse to hold all movie-related data in one response.
- Introduced MovieDetails domain model under the useCase.details package.
- Implemented MovieDetailRemoteMapper for mapping the combined remote response to the domain model.

**Use Case + Repository:**
- Updated GetMovieDetailsUseCase to return the new MovieDetails object.
- Refactored MovieRepositoryImpl and MovieRemoteDataSourceImpl to support the new unified fetching.
- Removed all obsolete fetching logic and mappers related to individual endpoints.

### TV Show Details
**API Integration:**
- Optimized TV show detail fetching via: /`tv/{tv_id}?append_to_response=credits,similar,reviews,images`
- Removed individual endpoints for credits, similar, reviews, and images from TvShowsApiService.

**Mapper + Domain Updates:**
- TvShowDetailsRemoteMapper now maps everything: cast, reviews, gallery, similar shows, production companies, and seasons using SeasonRemoteMapper.
- Updated TvShowDetailsRemoteResponse to include fields for the appended data.

---
## Checklist
Please ensure the following tasks are completed:

- [ ] My code follows the code style of this project
- [ ] Changes have been tested manually and verified.
- [ ] PR includes at most one single feature.

Belal-Alkhatib committed:
Refactor: Optimize movie details API calls

Belal-Alkhatib committed:
refactor: Streamline Movie Details Fetching and Mapping

Belal-Alkhatib committed:
refactor: optimize TV show details API call by using just one api call with append query

Belal-Alkhatib committed:
refactor: Remove unused imports and logging

Belal-Alkhatib committed:
Refactor: Move MovieDetails and TvShowDetails to UseCase

This commit refactors the codebase by moving the `MovieDetails` and `TvShowDetails` data classes from their respective domain models and remote mappers to their corresponding UseCase files (`GetMovieDetailsUseCase` and `GetTvShowDetailsUseCase`).

Additionally, the following changes were made:

- Simplified `GetTvShowDetailsUseCase` by directly returning the result from `tvShowRepository.getTvShowDetails(tvShowId)`.
- Corrected the `TvShowDetailsRemoteMapper` to use `SeasonRemoteMapper` for mapping seasons.
- Removed an unnecessary query parameter in `TvShowsApiService.getTvShowDetailsById`.

Belal-Alkhatib committed:
Merge branch 'develop' into feature/reduce-detail-requests

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/mapperModule.kt
#	domain/src/main/java/com/amsterdam/domain/useCase/details/GetTvShowDetailsUseCase.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImpl.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImpl.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/serviceProvider/implementation/MovieServiceProviderImpl.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/serviceProvider/implementation/TvShowsServiceProviderImpl.kt
#	repository/src/main/java/com/amsterdam/repository/dto/remote/TvShowDetailsRemoteResponse.kt
#	repository/src/main/java/com/amsterdam/repository/mapper/remote/MovieRemoteMapper.kt
#	repository/src/main/java/com/amsterdam/repository/mapper/remote/TvShowDetailsRemoteMapper.kt
#	repository/src/main/java/com/amsterdam/repository/mapper/remote/TvShowRemoteMapper.kt
#	repository/src/main/java/com/amsterdam/repository/repository/MovieRepositoryImpl.kt
#	repository/src/main/java/com/amsterdam/repository/repository/TvShowRepositoryImpl.kt

nabilBouzineDev commented on domain/src/main/java/com/amsterdam/domain/repository/TvShowRepository.kt:
```diff
@@ -1,22 +1,15 @@
     suspend fun getTopRatedTvShows(): List<TvShow>
     suspend fun getTvShowByKeyword(keyword: String, page: Int, tvShowsPerPage: Int): List<TvShow>
-    suspend fun getTvShowDetails(tvShowId: Long): TvShow
-    suspend fun getTvShowCast(tvShowId: Long): List<Actor>
```
@Belal-Alkhatib pay attention to this change because you deleted this function but it's used in `GetTvShowCastUseCase`

Belal-Alkhatib replied:
That was my oversight during the merge, I'll take care of it.

Belal-Alkhatib committed:
Refactor: Remove mapperModule and inject mappers directly

This commit removes the `mapperModule.kt` file, which was responsible for providing mapper instances through Koin. Mappers are now directly injected into their respective classes using constructor injection.

Additionally, the following changes were made:

-   Added `getTvShowCast` functionality to retrieve cast members for a TV show.
-   Integrated `addTvShowWatchHistoryUseCase` into `GetTvShowDetailsUseCase` to record watched TV shows.
-   Updated `MovieRepositoryImpl` and `TvShowRepositoryImpl` to reflect the new mapper injection strategy and incorporate the `getTvShowCast` functionality.
-   Adjusted `MovieDetailRemoteMapper` and `TvShowDetailsRemoteMapper` to use constructor injection.

Belal-Alkhatib committed:
Refactor: remove unused imports in repository and remoteDatasource modules

Belal-Alkhatib committed:
Merge branch 'develop' into feature/reduce-detail-requests

# Conflicts:
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImpl.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImpl.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/serviceProvider/MovieServiceProvider.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/serviceProvider/TvShowsServiceProvider.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/serviceProvider/implementation/MovieServiceProviderImpl.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/serviceProvider/implementation/TvShowsServiceProviderImpl.kt

saraashraf11 commented on domain/src/main/java/com/amsterdam/domain/repository/MovieRepository.kt:
```diff
@@ -13,13 +12,8 @@ interface MovieRepository {
-    suspend fun getMovieDetailsById(movieId : Long): Movie
+    suspend fun getMovieDetailsById(movieId : Long): MovieDetails
 
-    suspend fun getSimilarMovies(movieId : Long) : List<Movie>
```
Why did you delete them? I already use them.

Belal-Alkhatib replied:
where ?
